doc/dev: work on pull request guide

benesch opened the pull request.

ruchirK committed:
sql-parser: support ALTER INDEX ... {RESET,SET} logical_compaction_window

Closes #3636

Previously, users had to set `logical_compaction_window` through a command line
flag when running Materialize. This commit teaches Materialize how to set this
parameter on a per-index basis.

benesch committed:
doc/dev: add guide about submitting/reviewing PRs

justinj committed:
persistence: handle compacted topics

Fixes #4024.

This changes the behaviour of the persister so that instead of assuming
that the preceding message is always offset-1, to tag messages with
their predecessor.

TODO on tests.

Commit:
doc: remove kinesis alpha warning on kafka page

Commit:
Merge pull request #4099 from rjnn/doc

doc: remove kinesis alpha warning on kafka page

JLDLaughlin committed:
Fix unreachable panic in INSERT code (#4100)

Calls transform_query from handle_insert to fix the following
panic: "entered unreachable code: Expr::Nested not desugared".

JLDLaughlin committed:
Create filter_expr function (#4102)

Generates a sql_parser::ast::Expr from a ShowStatementFilter.

JLDLaughlin committed:
Order SHOW COLUMNS output in source order (#4101)

Adds an OrderByExpr to the Select statement generated in handle_show_columns.

ruchirK committed:
sql: rename normalize::with_options to normalize::options

ruchirK committed:
coord: rename *compaction_latency_ms to *compaction_window_ms

ruchirK committed:
testdrive: fix validate_sink and consolidation.td

Previously, validate_sink would not catch an error when comparing a single expected
record against a empty vec of actually received records, because the function would
zip two iterators together and one would be advanced, while the other was not. This
commit fixes that behavior by calling next() on both iterators in parallel and
complaining loudly if only one of them returns None.

test/testdrive/consolidation.td was broken in a few ways. First, the CREATE SOURCE
command and the kafka-ingest commands used incorrect Kafka topic names for the
consistency topic and the data topic respectively. In both cases, we forgot to
include the "testdrive-" prefix and "-${testdrive-seed}" suffix on topic names.

Furthermore, the timestamping code expects to get Debezium formatted consistency
information for ENVELOPE DEBEZIUM sources, and silently fails to parse the
"Materialize-native" format in that case.

dependabot-preview[bot] committed:
build(deps): bump async-trait from 0.1.38 to 0.1.40

Bumps [async-trait](https://github.com/dtolnay/async-trait) from 0.1.38 to 0.1.40.
- [Release notes](https://github.com/dtolnay/async-trait/releases)
- [Commits](https://github.com/dtolnay/async-trait/compare/0.1.38...0.1.40)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

ruchirK committed:
Merge pull request #4103 from MaterializeInc/dependabot/cargo/async-trait-0.1.40

build(deps): bump async-trait from 0.1.38 to 0.1.40

sploiselle committed:
demo/chbench: allow setting num warehouses

sploiselle committed:
demo/chbench: clean up kafka ports

madelynnblue committed:
doc: fix testing testing guide links

Commit:
Merge pull request #4108 from sploiselle/chbench-num-warehouses

demo/chbench benchmark tidying up

madelynnblue committed:
Merge pull request #4109 from mjibson/testing-links

doc: fix testing testing guide links

quodlibetor committed:
Update release checklist (#4110)

quodlibetor committed:
Update release checklist for new release blocker label

cirego committed:
Prepare v0.4.3 dev (#4111)

* release: v0.4.2-rc1

* Prepare next phase of development: v0.4.3-dev

quodlibetor committed:
Don't create an issue for the relnotes team

cirego committed:
Clarify how release notes are updated

Commit:
Intermediate step for 0.4.2 release notes

Collect all the relevant changes I could find.  Links directly to most PRs for reference (will be removed).

Commit:
Int64 schema_id

Commit:
Use proper github link syntax

First reviewable step for #4113

Commit:
link to appropriate docs

Commit:
fix broken urls + syntax highlight VALUES

JLDLaughlin committed:
Int64 schema_id (#4114)

Updates the schema_id field of the mz_catalog.mz_schemas system table
to be an Int64 instead of a String.

Commit:
Group memory-related changes together

Commit:
group memory changes together

Commit:
Merge pull request #4115 from MaterializeInc/awang-relnote-0.4.2

add relnotes for 0.4.2

benesch committed:
doc/user: tweak v0.4.2 release notes

Several fixes in here:

  * Move most release notes from v0.5.0 to v0.4.2. These release notes
    were only speculatively included in part of v0.5.0; now that v0.5.0
    is delayed, they will ship as v0.4.2 instead.

  * Remove release note about the `INSERT` panic. As a rule, we don't
    include release notes about experimental features, just to cut
    ourselves some slack.

  * Move release note about memory profiles out of the "improve
    memory utilization" section, as it does not improve memory
    utilization. There is probably a more general heading that would
    cover both memory usage and debugging improvements, but I'm not sure
    what it is, and this seemed easier.

  * Clarify and add links to several of the existing release notes to
    make them more relevant to end users.

  * Update the `SHOW COLUMNS` documentation to note the new ordering
    guarantees.

madelynnblue committed:
mzbuild: support specifying a build mode

madelynnblue committed:
testdrive: add some mzcompose env vars

ruchirK committed:
Merge pull request #4001 from ruchirK/rk-3636

sql-parser: support ALTER INDEX ... {RESET,SET} logical compaction window

ruchirK committed:
coord: augment source connectors with previously persisted files on-demand

Previously, we only checked to see if we could reuse already persisted data
after restarts, and we would then cache that information in the catalog. This
was a simple behavior but ultimately not very useful. This commit teaches the
coordinator to augment source connectors whenever they are being imported
into a dataflow, which opens up more opportunities for the persisted data to
be useful, and makes persistence easier to test.

ruchirK committed:
Merge pull request #4106 from ruchirK/rk-persistence-testdrive

coord: reuse persisted files more aggressively

quodlibetor committed:
Merge pull request #4112 from quodlibetor/rel-blockers

Update release checklist for new release blocker label

sploiselle committed:
demo/chbench: add environment logging for benchmark workflows

Commit:
Merge pull request #4126 from sploiselle/print-infra-env

demo/chbench: add environment logging for benchmark workflows

sploiselle committed:
demo/chbench: assert default number of threads through env var in benchmark

Commit:
Merge pull request #4128 from sploiselle/chbench-default-thread-fix

demo/chbench: assert default number of threads through env var

cirego committed:
Update testing instructions in release process (#4121)

Commit:
Remove long unused optimization. Fix FilterEqualLiteral so it compiles.

quodlibetor committed:
Update dashboard from grafana.mz (#4129)

This primarily fixes the metric name for messages behind kafka

Commit:
Update FilterEqualLiteral in light of Join having ScalarExpr equivalences.

JLDLaughlin committed:
Move mz_catalog system table docs (#4116)

Moves documentation for mz_catalog system tables (e.g., mz_catalog.mz_databases)
to a new system-tables.md document.

madelynnblue committed:
coord: add integer_datetimes server var

This variable has been hardcoded to on since postgres 10. It affects how
jdbc (and perhaps other drivers) interpret binary encoded timestamps. The
encoding we are using corresponds to the hardcoded on setting.

Fixes #4117

sploiselle committed:
demo/chbench: propagate NUM_WAREHOUSES to cloud-benchmark-infra

Commit:
Merge pull request #4133 from sploiselle/chbench-default-thread-fix

demo/chbench: propagate NUM_WAREHOUSES to cloud-benchmark-infra

sploiselle committed:
demo/chbench: rewrite q08 to stop erroring on division by zero

Commit:
demo/chbench: rewrite slt results

Commit:
Merge pull request #4134 from sploiselle/demo-chbench-q08-rewrite

chbench q08 rewrite

frankmcsherry committed:
Use monoids for monotonic min/max (#4122)

This PR uses differential dataflow's support for monoids (in place of integer counts) to accumulate min and max operators with monotonic (non-removing) inputs. This is appropriate for mins and maxes in a way that doesn't easily generalize to top k (at least, for non-trivial k), so trying it out here first.

There may be some iteration as errors are surfaced. In particular, this implementation will perform less type-based error checking than the prior implementation, which would panic on type errors (this implementation just goes with the smallest non-Datum::Null value, which could in principle be of different types if we have errors elsewhere in the system).

madelynnblue committed:
Merge pull request #4132 from mjibson/int-timestamps

coord: add integer_datetimes server var

umanwizard committed:
Random avro data in kgen (#4136)

umanwizard committed:
Fix "messages ingested" metric for upsert sources (#4137)

sploiselle committed:
docs: correct descriptions of string length function

Commit:
Merge pull request #4138 from sploiselle/str-length-desc

docs: correct descriptions of string length function

JLDLaughlin committed:
Remove qualified_name from mz_columns system table (#4139)

Replaces the information that was contained in the qualified_name field
with a join to the mz_catalog_names table.

justinj committed:
persistence: handle multiple readers

ruchirK committed:
dataflow: compact mz_scheduling_histogram and mz_scheduling_elapsed (#4127)

This commit uses a cyclic dataflow to compact the mz_scheduling_histogram logs and the mz_scheduling_elapsed logs and bring down the rate of memory growth under heavy query load in Materialize by only retaining log entries that correspond to currently active dataflows.

madelynnblue committed:
sql: add UUID type support

Add simple UUID type support with casting to and from strings.

The uuid crate appears well supported and comes with postgres
serialization support.

Fixes #3687

umanwizard committed:
Avro CDCv2 encoder/decoder (#4125)

JLDLaughlin committed:
ALTER RENAME statements update the mz_catalog_names system table (#4140)

When an object is updated via an ALTER RENAME statement, issue a retraction
for the old name and an addition of the new name to the mz_catalog.mz_catalog_names
system table.

umanwizard committed:
Use crate name "mz-avro" instead of "avro" internally (#4142)

benesch committed:
doc/user: correct requirements on SSL configuration

The ssl_ca_location is not actually required; if omitted, the system
default CA certificates are used.

benesch committed:
Merge pull request #4146 from benesch/ssl-ca

doc/user: correct requirements on SSL configuration

umanwizard committed:
Incrementally improve avro decoding API, by deriving simple cases (#4145)

Commit:
Fix typo; improve comments.

madelynnblue committed:
avro: add UUID support

Fixes #3577

umanwizard committed:
Allow deriving avro decoders that require state. (#4148)

madelynnblue committed:
docs: remove extra word

Commit:
Merge pull request #4149 from MaterializeInc/mjibson-patch-1

docs: remove extra word

madelynnblue committed:
Merge pull request #4143 from mjibson/uuid

sql: add UUID type support

cirego committed:
v0.4.2 has been released! (#4152)

* v0.4.2 has been released!

* Update config.toml

JLDLaughlin committed:
Add UUID type to build_row_schema_json match (#4154)

Fixes merge skew issue.

ruchirK committed:
test/perf-upsert: run lots of select count(*)s on mz

Touches #3234

frankmcsherry committed:
Lift logging retraction earlier (#4158)

This PR lifts the logging retraction action before count_total operators, which are stateful. It appears to result in a flat memory profile over time, though only evaluated on a few minute timescale. Thank to @ruchirK for the live coding session!

dependabot-preview[bot] committed:
build(deps): bump dogsdogsdogs from `a245417` to `db74efb`

Bumps [dogsdogsdogs](https://github.com/TimelyDataflow/differential-dataflow) from `a245417` to `db74efb`.
- [Release notes](https://github.com/TimelyDataflow/differential-dataflow/releases)
- [Commits](https://github.com/TimelyDataflow/differential-dataflow/compare/a245417fc24ee55bd640defdef3ea642fde8f34f...db74efb71a7644a9f0cddefc79657eba457bb16c)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

dependabot-preview[bot] committed:
build(deps): bump syn from 1.0.39 to 1.0.40

Bumps [syn](https://github.com/dtolnay/syn) from 1.0.39 to 1.0.40.
- [Release notes](https://github.com/dtolnay/syn/releases)
- [Commits](https://github.com/dtolnay/syn/compare/1.0.39...1.0.40)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

benesch committed:
Merge pull request #4160 from MaterializeInc/dependabot/cargo/syn-1.0.40

build(deps): bump syn from 1.0.39 to 1.0.40

benesch committed:
Merge pull request #4159 from MaterializeInc/dependabot/cargo/dogsdogsdogs-db74efb

build(deps): bump dogsdogsdogs from `a245417` to `db74efb`

dependabot-preview[bot] committed:
build(deps): bump proc-macro2 from 1.0.18 to 1.0.20

Bumps [proc-macro2](https://github.com/alexcrichton/proc-macro2) from 1.0.18 to 1.0.20.
- [Release notes](https://github.com/alexcrichton/proc-macro2/releases)
- [Commits](https://github.com/alexcrichton/proc-macro2/compare/1.0.18...1.0.20)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

ruchirK committed:
Merge pull request #4156 from ruchirK/rk-4155

test/perf-upsert: run lots of select count(*)s on mz

madelynnblue committed:
pgtest: add postgres wire protocol tester

Add a postgres wire protocol tester. This is useful because it allows
us to test things against a real postgres server and compare the results
to mz without needing to depend on drivers and their specific behavior.

To demonstrate this, a test file is included that demonstrates a current
bug in our pgwire implementation, hence why it isn't hooked up to
builds yet.

benesch committed:
Merge pull request #4161 from MaterializeInc/dependabot/cargo/proc-macro2-1.0.20

build(deps): bump proc-macro2 from 1.0.18 to 1.0.20

justinj committed:
Merge pull request #4079 from justinj/compacted-topics

persistence: handle compacted topics

JLDLaughlin committed:
Fix hanging sqllogictest runner (#4167)

Since I added the mz_catalog.mz_columns system table in #4085, the sqllogictest runner
has been hanging locally. Adding the mz_columns table meant that we were broadcasting
a new insert for every builtin column in a Materialize instance.

Because the sqllogictest runner tried to instantiate a Coordinator before it
launched its dataflow workers, the Coordinator was sending tons of messages into the
comm channel without anything to read them back out. This issue was fixed for actual
Materialize instances with #3284 by simply launching the dataflow workers first. However,
this fix was not applied to the sqllogictest runner at that time.

This PR updates the sqllogictest runner to launch its dataflow workers before it attempts
to instantiate the Coordinator, mirroring the fix in #3284.

benesch committed:
prof: simplify installation of malloc_conf

Export `malloc_conf` as a `&[u8]` rather than a `Option<&c_char>`. This
seems a good bit less magical.

Based on my read of the Rust reference's "Type Layout" chapter, pointers
and references have the same layout, and obviously a u8/i8s have the
same layout, so a symbol of type `&[u8]` should be exactly equivalent to
a symbol of type `*c_char`. The use of an option type in the old
approach actually strikes me as kind of sketchy; not sure what the
representational guarantees of `Option` are.

I've manually verified that this new approach works on Linux.

sploiselle committed:
peeker: add warmup and run duration parameters

sploiselle committed:
demo/chbench: add workflow to run prescribed chbench benchmark

benesch committed:
Merge pull request #4168 from benesch/malloc-conf-simple

prof: simplify installation of malloc_conf

Commit:
Database id should be an i64 in builtin tables

Commit:
Fix mz_databases columns in testdrive tests

JLDLaughlin committed:
Database id should be an i64 in builtin tables (#4173)

Updates database id representation in mz_databases and mz_schemas system tables
from a String to an Int64. Also updates the SHOW SCHEMAS query that references these
values.

Commit:
Rename catalog update functions

benesch committed:
*: use conventional capitalization for "Uuid"

Unlike in Go, the conventional capitalization of acronyms in Rust uses CamelCase. Adjust "UUID" to "Uuid" accordingly.

JLDLaughlin committed:
Rename catalog update functions (#4175)

Renames coordinator functions updating system tables to match the report_x_update
naming pattern.

madelynnblue committed:
Merge pull request #4119 from mjibson/testing

testing: improvements

JLDLaughlin committed:
Create new mz_indexes system table (#4130)

Adds a new mz_catalog.mz_indexes table that contains a unique row for each
index in a Materialize instance. Updates the implementation of handle_show_indexes
to query mz_indexes instead of building up index information from function calls
to the catalog.

benesch committed:
Merge pull request #4120 from benesch/042-relnotes

doc/user: tweak v0.4.2 release notes

Commit:
Merge pull request #4169 from sploiselle/peeker-durations

Add durations to peeker

frankmcsherry commented on doc/developer/guide-changes.md:
````diff
@@ -0,0 +1,605 @@
+
+In many cases, however, simplicity is clear cut. Don't write this:
+
+```rust
+fn is_cold(temp: i32) {
+    if temp < 62 {
+        true
+    } else {
+        false
+    }
+}
+```
+
+Write this:
+
+```rust
+fn is_cold(temp: i32) {
+    temp < 62
+}
+```
````
`fn is_cold(temp: i32) -> bool`!

benesch replied:
Oops, thanks!

umanwizard commented on doc/developer/guide-changes.md:
```diff
@@ -0,0 +1,605 @@
+the change is correct, and b) the change is uncontroversial. When in doubt, get
+a review!
+
+Most reviews are implicitly "+2", which means, "feel free to merge!" If a
```
Isn't this +2 vs. +1 stuff already covered by the distinction between "accepting" in the GitHub UI vs. just leaving comments?

I have always understood accept to mean you can merge it, comments to mean you can merge it if someone else approves, and reject to mean please don't merge it without me specifically signing off.

cuongdo replied:
I agree with @umanwizard. I have a soft preference for relying on GitHub's approve / comment actions to signal whether the review constitutes an approval -- the names of those actions make the intent easy to grasp.

benesch replied:
Ack, reworked this section accordingly.

umanwizard commented on doc/developer/guide-changes.md:
```diff
@@ -0,0 +1,605 @@
+
+Google's overriding principle is as follows:
+
+> Reviewers should favor approving a PR once it definitely improves the overall
```
You call them CLs in the other quotes from Google, but apparently changed it to PR here?

benesch replied:
Oh good point. Changed to "PR" everywhere.

umanwizard commented on doc/developer/guide-changes.md:
```diff
@@ -0,0 +1,605 @@
+
+There is a tricky distinction here that relies quite a bit on intuition. When
+you've nailed an API and you're adding a feature that fits the existing spirit
+of that API, the easy spot to add that feature will *also* be the right feature
```
The end of this sentence doesn't make sense to me, maybe `s/the right feature to add code/the right place to add it/` ?

benesch replied:
Done.

umanwizard commented on doc/developer/guide-changes.md:
```diff
@@ -0,0 +1,605 @@
+
+Our testing philosophy is covered in much greater detail in
+[Developer guide: testing](guide-testing.md).
+
```
I would add a paragraph like the following:

Note that whenever possible, tests should validate _observable behavior_, not _implementation details_. Over-testing of trivial internal functions creates its own maintenance burden, as tests frequently have to be changed because of innocuous refactors, causing people to be less likely to attempt refactors, or worse, less likely to take test failures seriously.

If a piece of code is large and complex enough, it might warrant tests to validate its behavior, despite not being externally observable. Even in this case, tests should validate the behavior at a clean API boundary of this code, rather than its internal details. As always, use your best judgment.

benesch replied:
Thanks, done.

madelynnblue committed:
Merge pull request #4165 from mjibson/pgtest

pgtest: add postgres wire protocol tester

Commit:
Have each delta of a DeltaQuery be on a separate line. Have each aggregation of a Reduce be on a separate line.

benesch committed:
Merge pull request #4177 from benesch/uuid-capitalization

*: use conventional capitalization for "Uuid"

benLogN committed:
Update broken link in README

benesch committed:
sql: handle dependent map exprs when simplifying quantified comparisons

The optimization that simplifies quantified comparisons was not properly
accounting for the fact that map expressions can refer to columns
introduced by earlier expressions in the same map node.

That map expressions work like this is the gift that just keeps on
giving.

Fix #4157.

umanwizard commented on doc/developer/guide-changes.md:
````diff
@@ -0,0 +1,605 @@
+they often start to grow extra parameters...
+
+```rust
+fn create_source(
````
```rust
fn open_file(
    fs: &mut Filesystem,
    uid: Option<u64>,
    gid: Option<u64>,
    path: String,
) -> Result<FileHandle, anyhow::Error> {
    if fs.type != "FAT32" && !(uid.is_some() && gid.is_some()) {
        bail!("Permissions checks are required on non-FAT32 systems!");
    } else {
        if (fs.type == "FAT32" || fs.type == "NTFS") && path.chars().any(|ch| ch == '/') {
            bail!("Windows paths use backslashes, not slashes!");
        } else {
            // Handle opening the file
            todo!()
        }
    }
}
```

benesch replied:
Done.

benesch committed:
Merge pull request #4183 from benesch/sql-subquery-panic

sql: handle dependent map exprs when simplifying quantified comparisons

Commit:
fix missing 'FROM' in docs

Commit:
Merge pull request #4184 from rjnn/fix_doc

fix missing 'FROM' in docs

wangandi committed:
Merge pull request #4178 from wangandi/cleanerexplain

EXPLAIN: Have each delta of a DeltaQuery and aggregation of a Reduce be on a separate line.

benesch committed:
sql: don't crash if parameters are used in a view

Prevent a query like

    CREATE VIEW v AS SELECT $1

from crashing Materialize. Now Materialize returns the same error
message as PostgreSQL ("there is no parameter $N").

Fix #3906.

madelynnblue committed:
Merge pull request #4166 from mjibson/uuid-avro

avro: add UUID support

umanwizard commented on doc/developer/guide-changes.md:
```diff
@@ -0,0 +1,611 @@
+[rust-rdkafka]: https://github.com/fede1024/rust-rdkafka
+[rust-prometheus]: https://github.com/MaterializeInc/rust-prometheus
+[rust-postgres]: https://github.com/sfackler/rust-postgres
+[source orthgonality]: https://github.com/MaterializeInc/materialize/issues/2537
```
stray

Commit:
SHOW DATABASES should work even if current database has been dropped

Fixes #4187

ruchirK committed:
dataflow: move persisted_files to live in KafkaSourceInfo

Commit:
Remove redundant test

JLDLaughlin committed:
SHOW DATABASES should work even if current database has been dropped (#4188)

Fixes unknown database error.

benesch committed:
Merge pull request #4185 from benesch/view-param-binding

sql: don't crash if parameters are used in a view

madelynnblue committed:
pgwire: correctly handle portal suspension/max rows logic

This PR fixes two bugs and now we match Postgres for the included test.

1) When max rows is specified and the number of remaining rows is
equal to max rows, we should return portal suspended instead of command
complete. Command complete should only be returned if more rows than
are remaining were returned.
2) If a portal has had all of its rows consumed, it should not be
re-executed. Instead it should return command complete with a count of 0.

madelynnblue committed:
pgtest: enable pgtests during CI

justinj committed:
sql: fix literal lifting through unions

Previously the logic for determining the literals to lift through a
union was wrong, because it special-cased the `base` input, and could
result in changing the shape of some of the inputs in an invalid way.

This change changes this behaviour to extract the longest common suffix
between all arms, which makes sure that the shape of the inputs is
always correct.

madelynnblue committed:
Merge pull request #4193 from mjibson/portals

pgwire: correctly handle portal suspension/max rows logic

umanwizard committed:
Make "unexpected foo" defaults in Avro decoders opt-in (#4194)

madelynnblue committed:
pgwire: move some tests to pgtest

madelynnblue committed:
pgtest: add some error options

Although not used yet, I used both of these doing testing and found
them useful.

ruchirK committed:
dataflow: read in files one at a time

dependabot-preview[bot] committed:
build(deps): bump libc from 0.2.76 to 0.2.77

Bumps [libc](https://github.com/rust-lang/libc) from 0.2.76 to 0.2.77.
- [Release notes](https://github.com/rust-lang/libc/releases)
- [Commits](https://github.com/rust-lang/libc/compare/0.2.76...0.2.77)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

dependabot-preview[bot] committed:
build(deps): bump differential-dataflow from `db74efb` to `2667ac2`

Bumps [differential-dataflow](https://github.com/TimelyDataflow/differential-dataflow) from `db74efb` to `2667ac2`.
- [Release notes](https://github.com/TimelyDataflow/differential-dataflow/releases)
- [Commits](https://github.com/TimelyDataflow/differential-dataflow/compare/db74efb71a7644a9f0cddefc79657eba457bb16c...2667ac284c2bb94d8074c028692a47f438760751)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

dependabot-preview[bot] committed:
build(deps): bump dogsdogsdogs from `db74efb` to `2667ac2`

Bumps [dogsdogsdogs](https://github.com/TimelyDataflow/differential-dataflow) from `db74efb` to `2667ac2`.
- [Release notes](https://github.com/TimelyDataflow/differential-dataflow/releases)
- [Commits](https://github.com/TimelyDataflow/differential-dataflow/compare/db74efb71a7644a9f0cddefc79657eba457bb16c...2667ac284c2bb94d8074c028692a47f438760751)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

umanwizard committed:
Avro typed errors [1/?] (#4198)

dependabot-preview[bot] committed:
build(deps): bump proc-macro2 from 1.0.20 to 1.0.21

Bumps [proc-macro2](https://github.com/alexcrichton/proc-macro2) from 1.0.20 to 1.0.21.
- [Release notes](https://github.com/alexcrichton/proc-macro2/releases)
- [Commits](https://github.com/alexcrichton/proc-macro2/compare/1.0.20...1.0.21)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

dependabot-preview[bot] committed:
build(deps): bump serde from 1.0.115 to 1.0.116

Bumps [serde](https://github.com/serde-rs/serde) from 1.0.115 to 1.0.116.
- [Release notes](https://github.com/serde-rs/serde/releases)
- [Commits](https://github.com/serde-rs/serde/compare/v1.0.115...v1.0.116)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

justinj committed:
Merge pull request #4195 from justinj/fix-lifting

sql: fix literal lifting through unions

benLogN committed:
Merge pull request #4179 from benLogN/cloud-link-fix

Update broken link in README

umanwizard committed:
Get rid of some dubious transmutes (#4199)

wangandi committed:
Merge pull request #4131 from wangandi/useindexcleanup

Cleanup use_indexes.rs

sploiselle committed:
peeker: add benchmark query group + thread lever

ruchirK committed:
dataflow,sql: regenerate RT source persisted record timestamps on startup.

Also disable persistence for BYO sources

Commit:
Add mz_tables system table

Each unique row contains the GlobalId and encompassing schema id of a table
in a Materialize instance.

Commit:
Merge pull request #4208 from sploiselle/peeker-benchmark-all-queries

peeker: add benchmark query group + thread lever

benLogN committed:
release: v0.4.3-rc1

benLogN committed:
Prepare next phase of development: v0.4.4-dev

benLogN committed:
Merge pull request #4214 from benLogN/prepare-v0.4.4-dev

Prepare v0.4.4 dev

umanwizard committed:
fix minor typo (#4217)

Commit:
Update implementation of SHOW TABLES

SHOW TABLES now reads from the mz_tables system table. Additionally, it
errors if the EXTENDED keyword is provided, which did not work in the prior
implementation.

Commit:
Use string formatted query instead

ruchirK committed:
dataflow: change file name format for persisted files.

Previously, persisted files were named [start, end] where `start` and `end`
denoted the range of offsets stored in this file as 1-indexed Kafka offsets.

This was downright confusing because Kafka offsets are actually 0-indexed, and
storing them in a 1-indexed format is a (hopefully temporary) Materialize-ism.

This commit tries to preserve as much existing behavior as possible while making
the generated file names for persisted files more understandable to the end user.
Now, file names have [start, end) where both are 0-indexed offsets, and `end` is the
next offset after the last one stored in the file.

madelynnblue committed:
Merge pull request #4196 from mjibson/pgtest-enable

pgtest: enable in CI; add some tests

sploiselle committed:
peeker: add olap_thread env var to loadtest group

madelynnblue committed:
pgwire: address comments from previous pgtest PR

Commit:
Merge pull request #4220 from sploiselle/loadtest-thread-count

peeker: add olap_thread env var to loadtest group

Commit:
Tidying

madelynnblue committed:
Merge pull request #4221 from mjibson/pgtest-comments

pgwire: address comments from previous pgtest PR

umanwizard committed:
Remove unnecessary annotation (#4222)

Commit:
SHOW EXTENDED TABLES not yet supported test

ruchirK committed:
Merge pull request #4209 from ruchirK/rk-persistence-load-test2

Change how Materialize reads in persisted data

Commit:
Remove GlobalIds from HashSet

ruchirK committed:
docs: add documentation for source persistence

benesch committed:
sql: support prepared INSERT statements

Fix #3941.

benesch committed:
Merge pull request #4204 from MaterializeInc/dependabot/cargo/proc-macro2-1.0.21

build(deps): bump proc-macro2 from 1.0.20 to 1.0.21

benesch committed:
Merge pull request #4203 from MaterializeInc/dependabot/cargo/dogsdogsdogs-2667ac2

build(deps): bump dogsdogsdogs from `db74efb` to `2667ac2`

benesch committed:
Merge pull request #4201 from MaterializeInc/dependabot/cargo/differential-dataflow-2667ac2

build(deps): bump differential-dataflow from `db74efb` to `2667ac2`

benesch committed:
Merge pull request #4202 from MaterializeInc/dependabot/cargo/serde-1.0.116

build(deps): bump serde from 1.0.115 to 1.0.116

benesch committed:
Merge pull request #4200 from MaterializeInc/dependabot/cargo/libc-0.2.77

build(deps): bump libc from 0.2.76 to 0.2.77

cuongdo commented on doc/developer/guide-changes.md:
```diff
@@ -0,0 +1,634 @@
+many of our prospective customers run their proofs-of-concept (POCs) off these
+latest unstable builds, so it is important to keep the `main` branch reasonably
+stable. Any correctness or performance regressions on the `main` branch are
+considered very high priority to fix.
```
Maybe `very high priority` -> `urgent`?

benesch replied:
Done.

Commit:
Add name to mz_tables

benesch committed:
sql-parser: remove some unnecessary boxes

Unclear where these came from; probably just copypasta.

Commit:
Create SchemaSpecifier, update SHOW TABLES

Commit:
Rename mz_tables field

madelynnblue committed:
coord: generate and persist a permanent cluster id

This id can be used to uniquely identify a materialize cluster across
restarts or being moved among VMs or containers.

Expose a `mz_cluster_id()` SQL function to retrieve it.

benesch committed:
doc: add release notes for 0.4.3

JLDLaughlin committed:
Merge pull request #4218 from JLDLaughlin/mz-table

The mz_catalog.mz_tables system table contains a unique row for each table in a Materialize
instance. SHOW TABLES reads from mz_tables, instead of computing the results.

madelynnblue committed:
Merge pull request #4230 from mjibson/cluster-id

coord: generate and persist a permanent cluster id

benesch committed:
Merge pull request #4227 from benesch/insert-prepared

sql: support prepared INSERT statements

Commit:
mz_sources and mz_sinks

Commit:
Missing trailing newline

Commit:
Small changes

Commit:
Spell out full SHOW SOURCES queries

umanwizard committed:
Make SchemaCache cache schema parsing and resolution errors (#4211)

benesch committed:
sql: fix merge skew on boxing of SelectStatement::query

ruchirK committed:
Merge pull request #4235 from benesch/merge-skew

sql: fix merge skew on boxing of SelectStatement::query

ruchirK committed:
dataflow: lower log verbosity for unimplemented persistence function

ruchirK committed:
Merge pull request #4236 from ruchirK/rk-fix-persistence-log

dataflow: lower log verbosity for unimplemented persistence function

Commit:
Update unreachable value

JLDLaughlin committed:
Merge branch 'main' into mz-cont

Commit:
Remove dereference

JLDLaughlin committed:
Merge pull request #4233 from JLDLaughlin/mz-cont

Add mz_catalog.mz_sources and mz_catalog.mz_sinks system tables to track
sources and sinks. These new system tables are used to replace the implementation of 
SHOW SOURCES and SHOW SINKS.

cuongdo commented on doc/developer/guide-changes.md:
```diff
@@ -0,0 +1,634 @@
+
+Some PRs, especially large refactors, are nearly impossible to split into
+pieces. If you find yourself in this situation, have a conversation with your
+reviewers ahead of time to get their buy in.
```
For larger refactors, I think it makes sense for the engineer to chat with another engineer about their design -- maybe the intended reviewer for the change. This prevents potentially wasted work and an awkward review if the reviewer disagrees with the basic premise of the change.

benesch replied:
Ack, added a section about early feedback.

benesch committed:
Merge pull request #4232 from benesch/relnotes-043

doc: add release notes for 0.4.3

cuongdo commented on doc/developer/guide-changes.md:
```diff
@@ -0,0 +1,634 @@
+
+## Reviewing changes
+
+The goals of code review are twofold:
```
Is there a third purpose: To socialize changes with other engineers?

benesch replied:
Sure, done!

cuongdo commented on doc/developer/guide-changes.md:
```diff
@@ -0,0 +1,634 @@
+tradeoffs to manage, and code review is an important component in communicating
+these particulars to engineers who are entering that region of the codebase for
+the first time. (Ideally these points are also documented somewhere, but even
+when they are, it is unreleastic to expect new engineers to read and absorb all
```
nit: `unreleastic` -> `unrealistic`

benesch replied:
Thanks, fixed.

cuongdo commented on doc/developer/guide-changes.md:
```diff
@@ -0,0 +1,634 @@
+and `SELECT * FROM 'nope'` is not, but it has no idea what either of those
+statements mean.
+
+XXX still working out this paragraph. It's tempting to add logic to the
```
reminder that this `XXX` is still here

benesch replied:
Thanks, fleshed out.

cuongdo reviewed (commented):
Nice doc! This appears to address recurring themes in code reviews. I think this is ready for wider circulation. None of my comments are blocking.

benLogN committed:
Update docs site for the completed 0.4.3 release

Add a note to the release issue template for how to find the deploy
in buildkite before starting the homebrew build.

benesch committed:
dataflow: drop capabilities for static file sources when finished

This regressed in #3532, and static file sources were having their
capabilities downgrade only every few seconds. Restore the original
behavior of dropping the capability entirely after the file is read so
that there is no need to periodically downgrade the capability at all.

This makes joining a table against a static file source much lower
latency. Fix #3975.

benesch committed:
doc/user: linkify mz_cluster_id in release notes

JLDLaughlin committed:
Support ALTER TABLE .. RENAME (#4239)

benesch committed:
Merge pull request #4241 from benesch/doc-mz-cluster-id

doc/user: linkify mz_cluster_id in release notes

benesch committed:
Merge pull request #4240 from benesch/finish-file-sources

dataflow: drop capabilities for static file sources when finished

benesch committed:
sql: stabilize tables

All known issues with tables have been fixed, so stabilize `CREATE
TABLE` and `INSERT`!

h/t to @JLDLaughlin for doing all of the heavy lifting here.

ruchirK committed:
materialized: remove the --persistence flag

Make persistence conditional on --experimental

benesch committed:
Merge pull request #4246 from benesch/stable-tables

sql: stabilize tables

benLogN committed:
Merge pull request #4238 from benLogN/0.4.3-docs

Update docs site for v0.4.3

benesch committed:
demo,test: upgrade to Debezium v1.2

benesch committed:
Merge pull request #4247 from benesch/dbz1.2

demo,test: upgrade to Debezium v1.2

ruchirK committed:
Merge pull request #4224 from ruchirK/rk-persistence-docs

docs: add documentation for source persistence

Commit:
Insert catalog items into coordinator in two loops

justinj committed:
sql: fix bug in reduce elision

Previously, reduce elision assumed that aggregates besides Count, when
fed a single row, evaluated to their input, but this was wrong for
SumInt32, which must evaluate to an Int64 rather than an Int32.

JLDLaughlin committed:
Insert catalog items into coordinator in two loops (#4249)

Sources and indexes may be depended upon. On starting the coordinator, ensure that
these catalog item types are inserted before other catalog item types to avoid errors.

JLDLaughlin committed:
Make handle_show implementations uniform (#4243)

Updates handle_show_databases, handle_show_schemas, and handle_show_indexes to generate 
query strings instead of programmatically building a Select statement. This update allows
us to clean up unnecessary helper functions.

Additionally, rename the Source_or_view output column from SHOW INDEXES to On_name.

justinj committed:
Merge pull request #4250 from justinj/fix-reduce-elision

sql: fix bug in reduce elision

Commit:
Allow test-util kafka_client to create more than one topic
and messages with null payload.

dependabot-preview[bot] committed:
build(deps): bump syn from 1.0.40 to 1.0.41

Bumps [syn](https://github.com/dtolnay/syn) from 1.0.40 to 1.0.41.
- [Release notes](https://github.com/dtolnay/syn/releases)
- [Commits](https://github.com/dtolnay/syn/compare/1.0.40...1.0.41)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

dependabot-preview[bot] committed:
build(deps): bump hyper from 0.13.7 to 0.13.8

Bumps [hyper](https://github.com/hyperium/hyper) from 0.13.7 to 0.13.8.
- [Release notes](https://github.com/hyperium/hyper/releases)
- [Changelog](https://github.com/hyperium/hyper/blob/master/CHANGELOG.md)
- [Commits](https://github.com/hyperium/hyper/compare/v0.13.7...v0.13.8)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

dependabot-preview[bot] committed:
build(deps): bump stacker from 0.1.11 to 0.1.12

Bumps [stacker](https://github.com/rust-lang/stacker) from 0.1.11 to 0.1.12.
- [Release notes](https://github.com/rust-lang/stacker/releases)
- [Commits](https://github.com/rust-lang/stacker/compare/stacker-0.1.11...stacker-0.1.12)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

dependabot-preview[bot] committed:
build(deps): bump datadriven from 0.2.0 to 0.3.0

Bumps [datadriven](https://github.com/justinj/datadriven) from 0.2.0 to 0.3.0.
- [Release notes](https://github.com/justinj/datadriven/releases)
- [Commits](https://github.com/justinj/datadriven/commits)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

benesch committed:
Merge pull request #4255 from MaterializeInc/dependabot/cargo/datadriven-0.3.0

build(deps): bump datadriven from 0.2.0 to 0.3.0

benesch committed:
Merge pull request #4252 from MaterializeInc/dependabot/cargo/syn-1.0.41

build(deps): bump syn from 1.0.40 to 1.0.41

benesch committed:
Merge pull request #4254 from MaterializeInc/dependabot/cargo/stacker-0.1.12

build(deps): bump stacker from 0.1.11 to 0.1.12

wangandi committed:
Merge pull request #4251 from wangandi/multitopickafkaclient

test-util: have KafkaClient support multiple topics + null payload

benesch committed:
Merge pull request #4253 from MaterializeInc/dependabot/cargo/hyper-0.13.8

build(deps): bump hyper from 0.13.7 to 0.13.8

benesch committed:
rust-toolchain: upgrade to Rust v1.46

benesch committed:
ci-builder: fix cargo caching

Avoid forwarding the host's `~/.cargo` directory, because that can
sometimes result in using the host's toolchain rather than the
container's, defeating the whole point of this setup.

benesch committed:
ci-builder: pin versions of cargo tools

Otherwise builds are not reproducible.

benesch committed:
ci/test: fix `cargo miri test` flag passing

Fix #4256.

benesch committed:
ci-builder: rebuild stable and nightly images

This incorporates the updates made in the past few commits.

benesch committed:
Merge pull request #4260 from benesch/ci-builder-rebuild

rust-toolchain: upgrade to Rust v1.46

JLDLaughlin committed:
Add mz_views system table and update SHOW VIEWS (#4261)

Adds a new mz_catalog.mz_views table to track each view in a Materialize
instance. Updates SHOW VIEWS to read from the mz_views table directly.

This update has the side effect of removing the is_queryable output column
from SHOW FULL VIEWS.

umanwizard committed:
Fix td avro parsing (#4266)

umanwizard committed:
update libz-sys to v1.1.2 to fix build with new xcode (#4271)

JLDLaughlin committed:
Add BI demo note for Mac users (#4276)
